### PR TITLE
Initial support for function pointers

### PR DESCRIPTION
**Pull request by densh**
Fix https://github.com/scala-native/scala-native/issues/36

### CODE REVIEW

File: nativelib/src/main/scala/scala/scalanative/native/FunctionPtr.scala [hunk 1]
```diff
@@ -0,0 +1,1279 @@
...
+import scalanative.runtime.undefined
+import scala.reflect.ClassTag
+
+/** C-style function pointer. */
+sealed abstract class FunctionPtr
+
+object FunctionPtr {
+
+// ###sourceLocation(file: "/Users/Denys/.src/native/nativelib/src/main/scala/scala/scalanative/native/FunctionPtr.scala.gyb", line: 14)
+
```
Comment (kmizu) [inline nativelib/src/main/scala/scala/scalanative/native/FunctionPtr.scala]: > ```
> ###sourceLocation(file: "/Users/Denys/.src/native/nativelib/src/main/scala/scala/scalanative/native/FunctionPtr.scala.gyb", line: 14)

I think it's not good that line comments include user specific information such as `/Users/Denys`.  Such information should be removed.

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala [hunk 2]
```diff
@@ -1198,6 +1226,58 @@ abstract class NirCodeGen
...
+    def genFunPtrOp(app: Apply, code: Int, focus: Focus): Focus = {
+      code match {
+        case FUN_PTR_CALL =>
+          val Apply(sel @ Select(funp, _), allargsp) = app
+
+          val arity = FunctionPtrApply.indexOf(sel.symbol)
+
+          val fun    = genExpr(funp, focus)
+          val argsp  = allargsp.take(arity)
+          val ctsp   = allargsp.drop(arity)
```
Comment (sjrd) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala]:
```scala
val (argsp, ctsp) = allargs.splitAt(arity)
```
?

**Comment by sjrd** [comment]
Do the `fromFunctionN` work for arbitrary `scala.FunctionN`s? It seems to me that they only work for eta-expanded extern functions. If that's the case, it should be documented in the Scaladocs of the conversions, otherwise you'll get zillions of complaints ;)

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala (hunk 2 above)
Comment (densh) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala]: 👍

**Comment by densh** [comment]
@sjrd I plan to expand it to arbitrary non-capturing lambdas today. Eta-expanded references to extern functions are quite limiting usabability-wise.

File: nativelib/src/main/scala/scala/scalanative/native/FunctionPtr.scala (hunk 1 above)
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/native/FunctionPtr.scala]: That's an artifact of our current code generation tool that we borrowed from Swift (called gyb.) There is an issue to replace it with something better (https://github.com/scala-native/scala-native/issues/135) but that's out-of-scope for M3.

**Comment by densh** [comment]
Wow, travis build fails because this pr introduces massive compilation speed regression (3x slower!) that overflows the time budget we're given. That never happened before.

**Comment by densh** [comment]
I've isolated the regression to anonymous lazy definition support. Still not clear what exactly causes 3x slowdown still.

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala [hunk 3]
```diff
@@ -76,11 +76,24 @@ abstract class NirCodeGen
...
-    val curMethodSym = new util.ScopedVar[Symbol]
-    val curEnv       = new util.ScopedVar[Env]
-    val curThis      = new util.ScopedVar[Val]
+    private val curLocalInfo = new util.ScopedVar[CollectLocalInfo]
+    private val curClassSym  = new util.ScopedVar[Symbol]
+    private val curMethodSym = new util.ScopedVar[Symbol]
+    private val curEnv       = new util.ScopedVar[Env]
+    private val curThis      = new util.ScopedVar[Val]
+
+    private val lazyAnonDefs = mutable.Map.empty[Symbol, ClassDef]
```
Comment (sjrd) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala]: It seems like you don't clear this map between compilation units.

Comment (densh) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala]: Spot on! Mutability, my old friend, strikes again.

Comment (sjrd) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala]: Reminds me of https://github.com/scala-js/scala-js/issues/2222 and its fix https://github.com/scala-js/scala-js/commit/533da726f8aad590af2692107e901c7f9bc234ea :p

Comment (densh) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala]: Those kinds of bugs are fun. 😄

File: sandbox/Test.scala [hunk 4]
```diff
@@ -1,13 +1,15 @@
-import scalanative.native._
+import scalanative.native._, FunctionPtr._
```
Comment (densh) [inline sandbox/Test.scala]: @sjrd Any idea why implicits in FunctionPtr are not resolved automatically? How does this work in SJS for ThisFunctions?

Comment (sjrd) [inline sandbox/Test.scala]: `js.ThisFunctionN` extends `js.ThisFunction`, and the implicit conversions are in the companion object of `js.ThisFunction`.

If you want automatic resolution, you either need `FunctionPtrN` to extend `FunctionPtr`, or put the implicit conversion to `FunctionPtrN` in the companion of `FunctionPtrN`.

Comment (densh) [inline sandbox/Test.scala]: This fixed, thanks.

**Pull request merged**